### PR TITLE
feat(css): add basic css parser (rdp)

### PR DESCRIPTION
**Pull request by neuodev**
This merge request implements basic CSS parser using [recursive descent parser](https://en.wikipedia.org/wiki/Recursive_descent_parser) algorithm and [Backus-Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) to defined the CSS grammar. 

Here is a simple example on how the parser generate the AST: 
```rs
        let mut parser = CSS3Parser::new();
        let style_sheet = parser.parse(
            r#"
            /* Shoud skip comments */

            .header {
                width: 10px;
                font-size: 1rem;
            } 

            #nav {
                height: 200px;
            }

            @media screen, printer, all and (max-width: 600px) {
                .header {
                    display: flex;
                }

                body {
                    max-height: 100vh;
                }
            }


            @media all {}
        "#,
        );

        assert_eq!(
            style_sheet,
            CSSStyleSheet {
                css_rules: vec![
                    CSSRule::CSSStyleRule(CSSStyleRule {
                        selector: CSSSelector::from_str_slice(vec![".", "header"]),
                        style_declarations: vec![
                            CSSStyleDeclaration {
                                name: "width".to_string(),
                                value: "10px".to_string()
                            },
                            CSSStyleDeclaration {
                                name: "font-size".to_string(),
                                value: "1rem".to_string()
                            }
                        ],
                    }),
                    CSSRule::CSSStyleRule(CSSStyleRule {
                        selector: CSSSelector::from_str_slice(vec!["#", "nav"]),
                        style_declarations: vec![CSSStyleDeclaration {
                            name: "height".to_string(),
                            value: "200px".to_string()
                        }],
                    }),
                    CSSRule::CSSMeidaRule(CSSMeidaRule {
                        conditions: vec![
                            MediaCondition::Type(MediaType::Screen),
                            MediaCondition::Type(MediaType::Printer),
                            MediaCondition::Type(MediaType::All),
                            MediaCondition::Feature(MediaFeature {
                                name: "max-width".to_string(),
                                value: Some("600px".to_string())
                            })
                        ],
                        css_rules: vec![
                            CSSRule::CSSStyleRule(CSSStyleRule {
                                selector: CSSSelector::from_str_slice(vec![".", "header"]),
                                style_declarations: vec![CSSStyleDeclaration {
                                    name: "display".to_string(),
                                    value: "flex".to_string()
                                }],
                            }),
                            CSSRule::CSSStyleRule(CSSStyleRule {
                                selector: CSSSelector::from_str_slice(vec!["body"]),
                                style_declarations: vec![CSSStyleDeclaration {
                                    name: "max-height".to_string(),
                                    value: "100vh".to_string()
                                }],
                            })
                        ],
                    }),
                    CSSRule::CSSMeidaRule(CSSMeidaRule {
                        conditions: vec![MediaCondition::Type(MediaType::All)],
                        css_rules: Vec::new(),
                    })
                ]
            }
        )
```

### CODE REVIEW

**Comment by jaytaph** [comment]
@neuodev could you try and fix the clippy issues?

**Comment by jaytaph** [comment]
how does it deal with styling errors? (i don't know how normal css parsers are doing this)?

File: src/css3_parser/tokenizer.rs
Comment (emwalker) [inline src/css3_parser/tokenizer.rs:43]: I'm guessing the use of regular expressions in the inner loop parser to identify the tokens will need to be provisional, for performance reasons.  But that can be profiled later and addressed if needed.

**Review by emwalker — APPROVED** [review]
I like it!  I like that we get an AST back that we can make assertions about in tests.

I would not be too attached to this code, as it is might change as we run into challenges and figure new things out.  But it looks like a promising start.

File: src/css3_parser/parser.rs
Comment (emwalker) [inline src/css3_parser/parser.rs:45]: I'm thinking it would eventually be good to drop the "CSS" prefix, and do something like
```rust
use crate::gosub::css3::{Selector, StyleRule};
// ...
```

But I wouldn't attempt that change in this PR, since it implies larger changes as well.

File: src/css3_parser/parser.rs
Comment (emwalker) [inline src/css3_parser/parser.rs:441]: I love this test.  It would be nice if we had things up so that the testing of the HTML parsing was like this as well.

File: src/css3_parser/parser.rs
Comment (emwalker) [inline src/css3_parser/parser.rs:411]:
```suggestion
    fn parses_css() {
```

**Review by Kiyoshika — APPROVED** [review]
I think this is an awesome start and will be iterated a lot in the future. We'll figure out how to deal with and how to handle the errors later so I'm not bothered about the panics yet.

I do agree with @emwalker that we should drop the `CSS` prefix in the struct names at a later point, but overall great!

File: src/css3_parser/parser.rs
Comment (Kiyoshika) [inline src/css3_parser/parser.rs:70]: I just noticed this now, is this supposed to me `Meida` or `Media`?

Comment (jaytaph) [reply src/css3_parser/parser.rs:70]: Do we need to fix this before merge? Doesn't seem a blocker

**Review by jaytaph — APPROVED** [review]

File: src/css3_parser/parser.rs
Comment (emwalker) [reply src/css3_parser/parser.rs:70]: This is easy to fix later on.  But eventually we'll want to block on this kind of thing, once the rate of PRs slows down or there's more reviewers.

Comment (Kiyoshika) [reply src/css3_parser/parser.rs:70]: Agree, and since the CSS parser won’t even be used for a bit/in hot development, I don’t think it’s a big deal to fix right away, maybe just create an issue so we don’t forget

**Comment by jaytaph** [comment]
Thanks @neuodev for the pr!

**Pull request merged**
